src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -294,27 +296,88 @@ private static void AddIsECMABoundaryHelper(Dictionary<string, string[]> require
         }
 
         /// <summary>Emits the body of the Scan method override.</summary>
-        private static void EmitScan(IndentedTextWriter writer, RegexMethod rm)
+        private static (bool NeedsTryFind, bool NeedsTryMatch) EmitScan(IndentedTextWriter writer, RegexMethod rm)
         {
             bool rtl = (rm.Options & RegexOptions.RightToLeft) != 0;
+            bool needsTryFind = false, needsTryMatch = false;
+            RegexNode root = rm.Tree.Root.Child(0);
 
-            using (EmitBlock(writer, "while (TryFindNextPossibleStartingPosition(inputSpan))"))
+            // We can alway emit our most general purpose scan loop, but there are common situations we can easily check
+            // for where we can emit simpler/better code instead.
+            if (root.Kind is RegexNodeKind.Empty)
             {
-                if (rm.MatchTimeout != Timeout.Infinite)
+                // Emit a capture for the current position of length 0.  This is rare to see with a real-world pattern,
+                // but it's very common as part of exploring the source generator, because it's what you get when you
+                // start out with an empty pattern.
+                writer.WriteLine("// The pattern matches the empty string.");
+                writer.WriteLine($"int pos = base.runtextpos;");
+                writer.WriteLine($"base.Capture(0, pos, pos);");
+            }
+            else if (root.Kind is RegexNodeKind.Nothing)
+            {
+                // Emit nothing.  This is rare in production and not something to we need optimize for, but as with
+                // empty, it's helpful as a learning exposition tool.
+                writer.WriteLine("// The pattern never matches anything.");
+            }
+            else if (root.Kind is RegexNodeKind.Multi or RegexNodeKind.One or RegexNodeKind.Notone or RegexNodeKind.Set && (root.Options & RegexOptions.IgnoreCase) == 0)
+            {
+                // If the whole expression is just one or more characters, we can rely on the FindOptimizations spitting out
+                // an IndexOf that will find the exact sequence or not, and we don't need to do additional checking beyond that.
+                needsTryFind = true;
+                using (EmitBlock(writer, "if (TryFindNextPossibleStartingPosition(inputSpan))"))
```

Comment:
Don't need to be done now, but I suppose that for these cases we can even further optimize even more and not even emit TryFindNextPossibleStartingPosition and instead just call IndexOf/IndexOfAny directly? RegexNodeKind.Notone would still need to be a bit special though.

Reply:
We could, I just didn't want to duplicate all that logic.  At that point we're veering into the "emit everything into Scan rather than a helper method" territory.